docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+
+### OPTIONS
+
+This HTTP verb will yield the allowed methods 
+OPTIONS https://zulip.tabbott.net/api/v1/users
+
+will yield, as a HTTP header:
+
+`Allow: POST, GET`
+
+You can see this reflected in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py):
+
+    url(r'^users$', 'zerver.lib.rest.rest_dispatch',
+        {'GET': 'zerver.views.users.get_members_backend',
+         'POST': 'zerver.views.users.create_user_backend'}),
+
+In this way, the API is partially self-documenting.
+
+### Legacy endpoints are used by the web client
+
+Many of endpoints from our legacy JSON API are written without REST in
+mind. They are still largely used by the web client, and use POST.
+
+You can see them in [zproject/legacy_urls.py](https://github.com/zulip/zulip/blob/master/zproject/legacy_urls.py).
+
+### Webhook integrations may not be RESTful
+
+Webhook--Zulip endpoints that are called by other services, especially
+for integrations, have to conform to the service's request format. Thus,
+they may not use REST, and are likely to use either POST or GET without
+the REST discipline of idempotency.
+
+TODO: mention webhook decorator
+
+## Django calls rest_dispatch for REST endpoints, and authenticates
+
+For reqeusts that correspond to a REST url pattern, Django will call
+`rest_dispatch` (see [zerver/lib/rest.py](https://github.com/zulip/zulip/blob/master/zerver/lib/rest.py)). This method will
+authenticate the user, either through a session token from a cookie, or
+from an `email:api-key` string given via HTTP Basic Auth for API clients.
+
+It will then look up what HTTP verb was used (GET, POST, etc) to make
+the request, and then figure out which view to show from that.
+
+## The view will authorize the user, extract request variables, and validate them
+
+There are some special decorators we may use for a given view:
+``` py
+@require_realm_admin
+@has_request_variables
+def create_user_backend(request, user_profile, email=REQ(), password=REQ(),
+                        full_name=REQ(), short_name=REQ()):
+```
+
+`require_realm_admin` authorizes the given `user_profile` to make sure
+it belongs to a realm admin and thus has permission to create a user.
+
+If you'll notice, there's a special `REQ()` keyword argument function.
+The body of a request is expected to be in JSON format, so this is used
+in conjunction with the `has_request_variables` decorator to unpack the
+variables in the JSON string for use in the function. The implementation
+of `has_request_variables` is documented heavily in [zerver/lib/request.py](https://github.com/zulip/zulip/blob/master/zerver/lib/request.py))
+
+REQ also helps with request variable validation. For example:
+`msg_ids = REQ(validator=check_list(check_int))`
+will check that the `msg_ids` request variable has the format of a list
+of integers.
+
+See [zerver/lib/validator.py](https://github.com/zulip/zulip/blob/master/zerver/lib/validator.py) for more validators and their documentation.
+
```

Comment:
I am now mentioning actions.py, but I will also be writing a "how to write a view" page that will cover it in more detail.